src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -262,34 +262,44 @@
       (ring-response/status http-status)
       (ring-response/content-type "text/plain")))
 
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;;; Public
+(defn wrap-with-error-handling
+  "Middleware that wraps a JRuby request with some error handling to return
+  the appropriate http status codes, etc."
+  [f]
+  (fn [request]
+    (sling/try+
+      (f request)
+      (catch bad-request? e
+        (output-error request e 400))
+      (catch jruby-timeout? e
+        (output-error request e 503)))))
 
-(defn handle-request*
-  [request jruby-instance config]
-  (->> request
-    wrap-params-for-jruby
-    (as-jruby-request config)
-    clojure.walk/stringify-keys
-    make-request-mutable
-    (.handleRequest jruby-instance)
-    response->map))
+(defn wrap-with-jruby-instance
+  "Middleware fn that borrows a jruby instance from the `jruby-service` and makes
+  it available in the request as `:jruby-instance`"
+  [f jruby-service]
+  (fn [request]
+    (jruby/with-jruby-puppet jruby-instance jruby-service
+      (f (assoc request :jruby-instance jruby-instance)))))
 
-(defmacro with-error-handling
-  [request & body]
-  `(sling/try+
-     ~@body
-     (catch bad-request? e#
-       (output-error ~request e# 400))
-     (catch jruby-timeout? e#
-       (output-error ~request e# 503))))
+(defn jruby-request-handler
+  "Build a request handler fn that processes a request using a JRubyPuppet instance"
+  [config]
+  (fn [request]
+    (->> request
+      wrap-params-for-jruby
+      (as-jruby-request config)
+      clojure.walk/stringify-keys
+      make-request-mutable
+      (.handleRequest (:jruby-instance request))
+      response->map)))
```

Comment:
Add a function here that returns a one-arg fn, accepting a 'request', and doing whatever stuff you need it to do.